packages/graphql-server/src/plugins/useRedwoodDirective.ts

```diff
@@ -0,0 +1,166 @@
+import { Plugin } from '@envelop/types'
+import { getArgumentValues } from '@graphql-tools/utils'
+import {
+  defaultFieldResolver,
+  DirectiveNode,
+  GraphQLObjectType,
+  GraphQLResolveInfo,
+  GraphQLSchema,
+} from 'graphql'
+
+import { GlobalContext } from '../index'
+
+function isQueryOrMutation(info: GraphQLResolveInfo): boolean {
+  const { parentType } = info
+
+  return parentType.name === 'Query' || parentType.name === 'Mutation'
+}
+
+export const DIRECTIVE_REQUIRED_ERROR_MESSAGE =
+  'You must specify one of @requireAuth, @skipAuth or a custom directive'
+
+export interface DirectiveArgs<FieldType = any> {
+  context: GlobalContext
+  directiveNode?: DirectiveNode
+  getFieldValue: () => FieldType
+  argumentValues: Record<string, any>
+  roles: [string] | undefined
+  root: unknown
+  args: Record<string, unknown>
+  info: GraphQLResolveInfo
+}
+
+/**
+ * Write your directive logic inside this function.
+ *
+ * - Return your transformed value if you want to replace it. e.g. masking a field
+ * - Throw an error, if you want to stop executing e.g. not sufficient permissions
+ *
+ */
+export type RedwoodDirective<FieldType = any> = (
+  args: DirectiveArgs<FieldType>
+) => FieldType | Promise<void> | void
+
+export type PluginOptions = {
+  onExecute: RedwoodDirective
+  name: string
+}
+
+export function hasDirective(info: GraphQLResolveInfo): boolean {
+  try {
+    const { parentType, fieldName, schema } = info
+    const schemaType = schema.getType(parentType.name) as GraphQLObjectType
+    const field = schemaType.getFields()[fieldName]
+    const astNode = field.astNode
+    // if directives array exists, we check the length
+    // other wise false
+    return !!astNode?.directives?.length
```

Comment:
@dac09 Should we change this rule to be more aligned with "secure by default" and validate that:

* has directives
* has requireAuth or skipAuth

In other works, all queries and mutations have to have requireAuth or skipAuth.

This won't let them set a different validation/access directive, though.

Reply:
Done